### PR TITLE
Features/multi input date #159742926

### PR DESCRIPTION
**Pull request by akegan**
https://www.pivotaltracker.com/story/show/159742926

This ended up being more changes than expected:
TL;DR is that I updated the domain DOB to be [YYYY, MM, DD] and added validation to the household member field

Still TODO:

- [x] - Verify that e2e tests work
- [x] - Add unit tests for date validation

### CODE REVIEW

**Comment by akegan** [comment]
moved back to work in progress because of the household member validation issues

**Comment by akegan** [comment]
This is ready to review again! I am writing up the review instructions now.

It now has the added functionality of validation for household member required fields (salesforce already errors out if first/last name are missing or dob is missing/invalid)

**Comment by akegan** [comment]
There ended up being more changes in this PR than expected, so please check out the review instructions and give this some extra QA love when reviewing. 

I also think this is probably a 2 person review kinda story

File: app/javascript/utils/form/Field.js [hunk 1]
```diff
@@ -24,6 +24,31 @@ export const Field = ({ formApi, field, label, blockNote, errorMessage, children
   )
 }
 
+export const MultiDateField = ({ formApi, field, label, id, blockNote, errorMessage, children }) => {
```
Comment (fddayan) [inline app/javascript/utils/form/Field.js]: I feel this component should be in its own file, like `MultiDateField.js`

File: app/javascript/utils/form/errors.js [hunk 2]
```diff
@@ -1,9 +1,13 @@
 import React from 'react'
+import { includes } from 'lodash'
 
 import reactFormUtils from '~/utils/reactFormUtils'
 
 export const hasError = (formApi, field) => {
-  if (reactFormUtils.get(formApi.touched, field) || formApi.submits > 0) { return !!formApi.getError(field) } else { return null }
+  var touched = reactFormUtils.get(formApi.touched, field)
```
Comment (fddayan) [inline app/javascript/utils/form/errors.js:7]: I would use `const`  and not `var`

**Comment by fddayan** [comment]
The input allows to enter no numerical values. Should we add `type=number` to the input to limit to only numbers?

image

File: app/javascript/utils/form/errors.js (hunk 2 above)
Comment (akegan) [reply app/javascript/utils/form/errors.js:7]: I'm changing the value on line 9

File: app/javascript/utils/form/Field.js (hunk 1 above)
Comment (akegan) [reply app/javascript/utils/form/Field.js]: done!

**Comment by akegan** [comment]
@fddayan I've move MultiDateField to a new file and specified type 'number' on the date input fields.

**Comment by callaghanc** [comment]
@akegan Since you mentioned above you thought this is a 2-person review story, I'm beginning a review on it as well now.

File: app/javascript/components/applications/application_form/PaperApplicationForm.js [hunk 3]
```diff
@@ -59,10 +60,24 @@ const buildPrefValidations = (prefs) => {
...
+
 const validateError = (values) => {
-  return {
-    preferences: buildPrefValidations(values.preferences)
+  const validateErrors = {
+    preferences: buildPrefValidations(values.preferences),
+    household_members: buildHouseholdMemberValidations(values.household_members)
   }
+  return validateErrors
 }
```
Comment (callaghanc) [inline app/javascript/components/applications/application_form/PaperApplicationForm.js]: trying out my first GitHub suggestion! ✨ looks like only one-line suggestions are supported at this time, so i wasn't able to select the entire `validateError` function, but hopefully you can see my suggestion is for a slight refactor to that function, not to replace just this bracket lol
```suggestion
const validateError = (values) => (
  {
    preferences: buildPrefValidations(values.preferences),
    household_members: buildHouseholdMemberValidations(values.household_members)
  }
)
```

File: app/javascript/components/mappers/soqlToDomain/application_member.js [hunk 4]
```diff
@@ -29,3 +29,10 @@ export const mapApplicationMember = (value) => {
     alternate_contact_type_other: value.Alternate_Contact_Type_Other
   }
 }
+
+const dateOfBirthToDomain = (dateOfBirth) => {
+  // Convert 'YYYY-MM-DD' string to domain format [YYYY, MM, DD]
+  if (!dateOfBirth) return null
+  const split = dateOfBirth.split('-')
```
Comment (callaghanc) [inline app/javascript/components/mappers/soqlToDomain/application_member.js]: maybe a more semantic name here for `split`? sth like `dateParts`?

File: app/javascript/utils/form/MultiDateField.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+          field={[field, 1]}
+          id={id + '_month'}
+          placeholder='MM'
+          type='number' />
+      </div>
+      <div className='form-group-day'>
+        <Text
+          className={classNames(className)}
+          field={[field, 2]}
+          id={id + '_day'} placeholder='DD'
```
Comment (callaghanc) [inline app/javascript/utils/form/MultiDateField.js]:
```suggestion
          id={id + '_day'}
          placeholder='DD'
```

File: app/javascript/utils/form/MultiDateField.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import React from 'react'
+import classNames from 'classnames'
+import { Text } from 'react-form'
+import { BlockNote } from '~/utils/form/Field'
+import { FormError, errorClassName } from './errors'
+
+export const MultiDateField = ({ formApi, field, label, id, blockNote, errorMessage, children }) => {
```
Comment (callaghanc) [inline app/javascript/utils/form/MultiDateField.js]: looks like we're not using `children` in here - could we remove it from the args?

File: spec/javascript/utils/form/validations.test.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import validate from 'utils/form/validations'
+
+// Validations will return this message if failed, null if passed
+const FAILED_VALIDATION = 'failed validation'
```
Comment (callaghanc) [inline spec/javascript/utils/form/validations.test.js]: could we update this name to `FAILED_VALIDATION_MSG`? that way it's easier to read that it's not a boolean validation result or an object or field that failed validation

File: spec/javascript/utils/form/validations.test.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+describe('validate', () => {
+  describe('isOldEnough', () => {
+    test('passes validation if DOB is more than 18 years ago', () => {
+      expect(validate.isOldEnough(FAILED_VALIDATION)(['2000', '01', '12'])).toEqual(null)
+    })
+    test('fails validation if DOB is less than 18 years ago', () => {
```
Comment (callaghanc) [inline spec/javascript/utils/form/validations.test.js:13]: one hopes this test will be updated before 2028, but to be proper about it, perhaps we could use moment to dynamically provide a date that is <18 years prior to the current date? not a priority obv

File: spec/javascript/utils/form/validations.test.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    test('passes validation if date is valid and not 0-padded', () => {
+      expect(validate.isValidDate(FAILED_VALIDATION)(['2000', '2', '9'])).toEqual(null)
+    })
+    test('passes validation if date is valid and values are integers', () => {
+      expect(validate.isValidDate(FAILED_VALIDATION)([2000, 2, 9])).toEqual(null)
+    })
+    test('fails validation if date is a leap day in the wrong year', () => {
+      expect(validate.isValidDate(FAILED_VALIDATION)(['1999', '02', '29'])).toEqual(FAILED_VALIDATION)
+    })
+    test('fails validation if date is in the future', () => {
```
Comment (callaghanc) [inline spec/javascript/utils/form/validations.test.js:34]: same comment as above about being excessively proper and using dynamically generated dates

**Review by callaghanc — CHANGES_REQUESTED** [review]
Looks good! I had some minor change suggestions, nothing big.

File: spec/javascript/utils/form/validations.test.js (hunk 8 above)
Comment (akegan) [reply spec/javascript/utils/form/validations.test.js:13]: fixed

File: spec/javascript/utils/form/validations.test.js (hunk 9 above)
Comment (akegan) [reply spec/javascript/utils/form/validations.test.js:34]: fixed

**Comment by akegan** [comment]
@callaghanc ready for review again!

**Review by callaghanc — APPROVED** [review]

**Pull request merged**